### PR TITLE
Switch Windows embedding to proc table embedder API

### PR DESCRIPTION
**Pull request by stuartmorgan**
## Description

Switches the Windows embedding from the standard C API to the new proctable version, to allow for unit testing of the embedding layer separately from the embedder APIs implementation. This includes moving some engine messaging that was still in flutter_windows to the C++ engine class to better encapsulate the proc table.

## Related Issues

https://github.com/flutter/flutter/issues/66935

## Tests

I added the following tests: Added a simple messaging unit test as a demonstration/proof of concept.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [ ] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by stuartmorgan** [comment]
@cbracken @gw280 Ping

File: shell/platform/windows/testing/engine_embedder_api_modifier.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/windows/flutter_windows_engine.h"
+
+namespace flutter {
+
+// A test utility class providing the ability to access and alter the embedder
+// API proc table for an engine instance.
```
Comment (gw280) [inline shell/platform/windows/testing/engine_embedder_api_modifier.h:10]: nit: Can you add a comment here saying that any modifications will outlive the lifetime of this object?

File: shell/platform/windows/flutter_windows_engine.cc [hunk 2]
```diff
@@ -232,6 +237,58 @@ void FlutterWindowsEngine::SetPluginRegistrarDestructionCallback(
...
 
+void FlutterWindowsEngine::SendWindowMetricsEvent(
+    const FlutterWindowMetricsEvent& event) {
+  if (engine_) {
+    embedder_api_.SendWindowMetricsEvent(engine_, &event);
+  }
+}
+
+void FlutterWindowsEngine::SendPointerEvent(const FlutterPointerEvent& event) {
+  embedder_api_.SendPointerEvent(engine_, &event, 1);
```
Comment (gw280) [inline shell/platform/windows/flutter_windows_engine.cc:248]: do we need to null-check engine_ here? If not, is the one in SendWindowMetricsEvent necessary?

File: shell/platform/windows/flutter_windows_engine.h [hunk 3]
```diff
@@ -79,11 +78,37 @@ class FlutterWindowsEngine {
...
   }
 
+  // Informs the engine that the window metrics have changed.
+  void SendWindowMetricsEvent(const FlutterWindowMetricsEvent& event);
+
+  // Informs the engine of an incoming pointer event.
+  void SendPointerEvent(const FlutterPointerEvent& event);
+
+  // Sends the given message to the engine, calling |reply| with |user_data|
+  // when a reponse is recieved from the engine if they are non-null.
```
Comment (gw280) [inline shell/platform/windows/flutter_windows_engine.h:88]: nit: typo in received

**Review by gw280 — APPROVED** [review]
some minor nits but looks good otherwise. In general though this code is hard to follow and we might want to file an issue to clean it up a little bit.

File: shell/platform/windows/testing/engine_embedder_api_modifier.h (hunk 1 above)
Comment (stuartmorgan) [reply shell/platform/windows/testing/engine_embedder_api_modifier.h:10]: Done; added both here and to embedder_api() just to make sure it's hard to miss since given the oddity of the API.

File: shell/platform/windows/flutter_windows_engine.cc (hunk 2 above)
Comment (stuartmorgan) [reply shell/platform/windows/flutter_windows_engine.cc]: Couldn't hurt; I couldn't easily convince myself that we couldn't have races around pointer event handling during teardown. Added.

File: shell/platform/windows/flutter_windows_engine.h (hunk 3 above)
Comment (stuartmorgan) [reply shell/platform/windows/flutter_windows_engine.h]: Done.
